### PR TITLE
Merge branch development to branch master

### PR DESCRIPTION
**Pull request by karine2711**

### CODE REVIEW

File: src/exception/DefaultHandler.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                    new Student.StudentBuilder("Dallas Knight", defaultCourseList).build(),
+            },
+            {
+                    new Student.StudentBuilder("Tanya Peterson", defaultCourseList).build(),
+                    new Student.StudentBuilder("Clifton Carter", "Data Structures",
+                            "Theory of Algorithms", "Intro to CS", "Business").build(),
+            },
+            {
+                    new Student.StudentBuilder("Mildred Nelson", defaultCourseList).build(),
+                    new Student.StudentBuilder("Pamela Norton", "Data Structures",
```
Comment (DiannaAsatryan) [inline src/exception/DefaultHandler.java]: Builder pattern used incorrectly, see this post: 
https://dzone.com/articles/design-patterns-the-builder-pattern
The idea is to have many optional fields which you will give to Builder with separate method then build your original object in builder

File: src/exception/CourseGrade.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package exception;
```
Comment (DiannaAsatryan) [inline src/exception/CourseGrade.java]: change package name to be task name like : 'university'

File: src/exception/DynamicArray.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
+package exception;
+
+public class DynamicArray {
```
Comment (DiannaAsatryan) [inline src/exception/DynamicArray.java]: It is not right to have a common class that works with most of your classes, and I see it is out of scope for this task to dynamically increase or decrease students,Faculty, Groups.  ... .You can just create static one, or initialize it in main and use after that.
If you need to have a class which does dynamic array length changes, you should just have DynamicArray<T> and have methods (non static) to work with T type, Like it done in ArrayList

File: src/exception/Student.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        return fullName;
+    }
+
+    public void setGrade(String courseName, double grade) {
+        int index;
+        index = UniversityService.search(this.courseGrades, courseName);
+        if (index == -1) {
+            System.out.println("The student doesn't study course named " + courseName);
+        } else if (grade <= 0 || grade > 10) {
+            System.out.println("The grade must be in the range (0,10]. Given grade " + grade);
```
Comment (DiannaAsatryan) [inline src/exception/Student.java]: Solution is very complicated, and when I run it throws NulPointerException  after many tries to exit from dynamic creation process, pls change just to have static initialized objects , and I don't see Exception handling at all, for example in his case it is expected to throw some custom created exception without doing sout, and catch it in caller method

File: src/exception/Main.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        //try to get a student and set a grade
+        try {
+            Student karine = UniversityService.getStudentByInfo(
+                    university,
+                    "CS",
+                    "A",
+                    "Karine Gevorgyan"
+            );
+            karine.setGrade("Armenian Literature", 5);
+        } catch (NullPointerException e) {
```
Comment (DiannaAsatryan) [inline src/exception/Main.java]: Don't catch NullPointerException, then create your own Exception which will extend RunTimeException and throw it getStudentByInfo(..) method like NoStudentException

File: src/exception/DefaultHandler.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                    new Group("B", defaultStudentList[1]),
+            },
+            {
+                    new Group("A", defaultStudentList[2]),
+                    new Group("B", defaultStudentList[3]),
+            },
+
+    };
+
+    private static int i = 0;
```
Comment (DiannaAsatryan) [inline src/exception/DefaultHandler.java]: not good to have fields which you can't guess what they are doing

**Review by DiannaAsatryan — CHANGES_REQUESTED** [review]
Karine jan pls make it simple, with static initialization of your objects and use Exception Handlings, and Bilder correctly, then I will do one more review

File: src/exception/DefaultHandler.java (hunk 1 above)
Comment (karine2711) [reply src/exception/DefaultHandler.java]: Could you please give a hint what exactly can be changed here?

File: src/university/Main.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+
+        //  count mean grade in a given group, for a given subject
+        try {
+            System.out.println("------------------------------------------------------");
+            System.out.println("--Count mean grade in a given group, for a given subject--");
+            Group group = UniversityService.getGroupByInfo(university, "EC", "EC1");
+            double grade = group.countMeanGrade("Music");
+            if (grade == 0) {
+                System.out.println("No one in the group passes the given course, or the students were not graded!");
```
Comment (DiannaAsatryan) [inline src/university/Main.java]: Karine  jan you have different splited codes by logic, why don't move each part to some small methods , for example: 
void printMinGradeOfGroupForSubject(University university, String groupName, String subject){
try {
            System.out.println("------------------------------------------------------");
            System.out.println("--Count mean grade in a given group, for a given subject--");
            Group group = UniversityService.getGroupByInfo(university, groupName,  subject);
            double grade = group.countMeanGrade("Music");
            if (grade == 0) {
                System.out.println("No one in the group passes the given course, or the students were not graded!");
 
} else {
                System.out.println("Mean grade for given subject in group "
                        + group.getName()
                        + " = " + grade
                );
            }
        } catch (ItemNotFoundException e) {
            System.out.println(e.toString());
        } catch (NullListException e) {
            e.printStackTrace();
        }
 

and the same for other parts

File: src/university/exceptions/GradeOutOfRangeException.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+    public GradeOutOfRangeException(double grade,int minGrade,int maxGrade){
+        this.grade=grade;
+        this.minGrade=minGrade;
+        this.maxGrade=maxGrade;
+    }
+
+    @Override
+    public String toString() {
+        return "Grade Out Of Range! Grade must be in range (" +
```
Comment (DiannaAsatryan) [inline src/university/exceptions/GradeOutOfRangeException.java]: as I understand you learned to create exceptions by overriding toString() in class, but what I will suggest calling a constructor att creation time and providing some message to it like :
 public GradeOutOfRangeException(double grade,int minGrade,int maxGrade){
super(String.format("Grade Out Of Range! Grade must be in range [%s,%s]       Given grade %s",minGrade,maxGrade,grade));
    }

 @Override
    public String toString() {
this.getMessage();
}

as in most of the cases when you use logger, we use e.getMessage()

File: src/university/universityParts/University.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @author Karine Gevorgyan
+ */
+
+public class University {
+
+    final static DecimalFormat numberFormat = new DecimalFormat("#.0");
+    private String name;
+    private Faculty[] faculties;
+    private String[] facultyNames;
```
Comment (DiannaAsatryan) [inline src/university/universityParts/University.java]: Karine jan don't store  facultyNames in class if you calculate them each time with getFacultyNames(..) method

File: src/university/universityParts/University.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
+package university.universityParts;
+
+import university.exceptions.nullListExceptions.NullFacultyListException;
+
+import java.text.DecimalFormat;
+
+/**
+ * University is a singletone class, as the program is written for a specific university.
```
Comment (DiannaAsatryan) [inline src/university/universityParts/University.java:8]: Karine jan your University is not singleton class as you have public constructor, remove these docs or make it singleton

File: src/university/universityParts/Student.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    final static int MIN_GRADE = 0;
+    final static int MAX_GRADE = 10;
+    private String fullName;
+    private Course[] courses;
+    private String address;
+    private String phone;
+    private String email;
+
+    private Student(StudentBuilder student) {
+        this.fullName = student.fullName;
```
Comment (DiannaAsatryan) [inline src/university/universityParts/Student.java:31]: builder pattern used incorrectly, the idea is to fill all Student's fields from builder

File: src/university/exceptions/nullListExceptions/NullCourseListException.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package university.exceptions.nullListExceptions;
+
+public class NullCourseListException extends NullListException {
+    private String studentName = "";
+
+    public NullCourseListException(String studentName) {
```
Comment (DiannaAsatryan) [inline src/university/exceptions/nullListExceptions/NullCourseListException.java]: it is good that you used inheritance here

**Review by DiannaAsatryan — APPROVED** [review]
well done Karine jan , you can merge it

**Pull request merged**